File: src/main/java/svnserver/TemporaryOutputStream.java
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.TestOnly;

import java.io.*;
import java.nio.file.Files;
import java.nio.file.Path;

/**
 * Stream for write-then-read functionality.
 *
 * @author Artem V. Navrotskiy
 * @author Marat Radchenko <marat@slonopotamus.org>
 */
public final class TemporaryOutputStream extends OutputStream {
  @SuppressWarnings("MagicNumber")
  public static final int MAX_MEMORY_SIZE = 8 * 1024 * 1024;

  private final int maxMemorySize;
  @NotNull
  private final ByteArrayOutputStream memoryStream = new ByteArrayOutputStream();
  @Nullable
  private Path file;
  @Nullable
  private OutputStream fileOutputStream;
  private long totalSize = 0;
  private boolean closed;

  public TemporaryOutputStream() {
    this(MAX_MEMORY_SIZE);
  }

  public TemporaryOutputStream(int maxMemorySize) {
    this.maxMemorySize = maxMemorySize;
  }

  @Override
  public void write(int b) throws IOException {
    if (closed)
      throw new IOException();

    if (memoryStream.size() < maxMemorySize) {
      memoryStream.write(b);
      totalSize++;
      return;
    }
    ensureFile().write(b);
    totalSize++;
  }

  @NotNull
  private OutputStream ensureFile() throws IOException {
    if (fileOutputStream == null) {
      file = Files.createTempFile("tmp", "git-as-svn");
      fileOutputStream = Files.newOutputStream(file);
    }
    return fileOutputStream;
  }

  @Override
  public void write(@NotNull byte[] b, int off, int len) throws IOException {
    if (closed)
      throw new IOException();

    if (memoryStream.size() < maxMemorySize) {
      final int size = Math.min(maxMemorySize - memoryStream.size(), len);
      memoryStream.write(b, off, size);
      if (size < len) {
        ensureFile().write(b, off + size, len - size);
      }
    } else {
      ensureFile().write(b, off, len);
    }
    totalSize += len;
  }

  @Override
  public void flush() throws IOException {
    if (closed)
      throw new IOException();

    if (fileOutputStream != null)
      fileOutputStream.flush();
  }

  @Override
  public void close() throws IOException {
    if (closed)
      return;

    closed = true;

    try {
      if (fileOutputStream != null)
        fileOutputStream.close();
    } finally {
      if (file != null)
        Files.deleteIfExists(file);
    }
  }

  public long size() {
    return totalSize;
  }

  @TestOnly
  @Nullable
  Path tempFile() {
    return file;
  }

  @NotNull
  public InputStream toInputStream() throws IOException {
    if (closed)
      throw new IOException();

    if (fileOutputStream != null)
      flush();

    final InputStream result = file == null
        ? new ByteArrayInputStream(memoryStream.toByteArray())
        : new TemporaryInputStream(memoryStream.toByteArray(), file);

    file = null;
    close();

    return result;
  }

  private static class TemporaryInputStream extends InputStream {
    @NotNull
    private final byte[] memoryBytes;
    @NotNull
    private final InputStream fileStream;
    @NotNull
    private final Path file;
    private int offset = 0;

    private TemporaryInputStream(@NotNull byte[] memoryBytes, @NotNull Path file) throws IOException {
      this.memoryBytes = memoryBytes;
      this.fileStream = Files.newInputStream(file);
      this.file = file;
    }

    @Override
    public int read() throws IOException {
      if (offset < memoryBytes.length) {
        //noinspection MagicNumber
        return memoryBytes[offset++] & 0xff;
      }
      return fileStream.read();
    }

    @Override
    public int read(@NotNull byte[] buf, int off, int len) throws IOException {
      if (len == 0) {
        return 0;
      }
      if (this.offset < memoryBytes.length) {
        final int count = Math.min(len, memoryBytes.length - this.offset);
        System.arraycopy(memoryBytes, offset, buf, off, count);
        offset += count;
        return count;
      }
      return fileStream.read(buf, off, len);
    }

    @Override
    public void close() throws IOException {
      try {
        fileStream.close();
      } finally {
        Files.deleteIfExists(file);
      }
    }
  }
}

File: src/main/kotlin/svnserver/TemporaryOutputStream.kt
/*
 * This file is part of git-as-svn. It is subject to the license terms
 * in the LICENSE file found in the top-level directory of this distribution
 * and at http://www.gnu.org/licenses/gpl-2.0.html. No part of git-as-svn,
 * including this file, may be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file.
 */
package svnserver

import org.jetbrains.annotations.TestOnly
import java.io.*
import java.nio.file.Files
import java.nio.file.Path
import kotlin.math.min

/**
 * Stream for write-then-read functionality.
 *
 * @author Artem V. Navrotskiy
 * @author Marat Radchenko <marat@slonopotamus.org>
 */
class TemporaryOutputStream @JvmOverloads constructor(private val maxMemorySize: Int = MAX_MEMORY_SIZE) : OutputStream() {
    private val memoryStream: ByteArrayOutputStream = ByteArrayOutputStream()
    private var file: Path? = null
    private var fileOutputStream: OutputStream? = null
    private var totalSize: Long = 0
    private var closed: Boolean = false

    @Throws(IOException::class)
    override fun write(b: Int) {
        if (closed) throw IOException()
        if (memoryStream.size() < maxMemorySize) {
            memoryStream.write(b)
            totalSize++
            return
        }
        ensureFile().write(b)
        totalSize++
    }

    @Throws(IOException::class)
    private fun ensureFile(): OutputStream {
        if (fileOutputStream == null) {
            file = Files.createTempFile("tmp", "git-as-svn")
            fileOutputStream = Files.newOutputStream(file)
        }
        return (fileOutputStream)!!
    }

    @Throws(IOException::class)
    override fun write(b: ByteArray, off: Int, len: Int) {
        if (closed) throw IOException()
        if (memoryStream.size() < maxMemorySize) {
            val size: Int = min(maxMemorySize - memoryStream.size(), len)
            memoryStream.write(b, off, size)
            if (size < len) {
                ensureFile().write(b, off + size, len - size)
            }
        } else {
            ensureFile().write(b, off, len)
        }
        totalSize += len.toLong()
    }

    @Throws(IOException::class)
    override fun flush() {
        if (closed) throw IOException()
        fileOutputStream?.flush()
    }

    @Throws(IOException::class)
    override fun close() {
        if (closed) return
        closed = true
        try {
            fileOutputStream?.close()
        } finally {
            if (file != null) Files.deleteIfExists(file)
        }
    }

    fun size(): Long {
        return totalSize
    }

    @TestOnly
    fun tempFile(): Path? {
        return file
    }

    @Throws(IOException::class)
    fun toInputStream(): InputStream {
        if (closed) throw IOException()
        if (fileOutputStream != null) flush()
        val result: InputStream = if (file == null) ByteArrayInputStream(memoryStream.toByteArray()) else TemporaryInputStream(memoryStream.toByteArray(), file!!)
        file = null
        close()
        return result
    }

    private class TemporaryInputStream(private val memoryBytes: ByteArray, private val file: Path) : InputStream() {
        private val fileStream: InputStream = Files.newInputStream(file)
        private var offset: Int = 0

        @Throws(IOException::class)
        override fun read(): Int {
            if (offset < memoryBytes.size) {
                return memoryBytes[offset++].toInt() and 0xff
            }
            return fileStream.read()
        }

        @Throws(IOException::class)
        override fun read(buf: ByteArray, off: Int, len: Int): Int {
            if (len == 0) {
                return 0
            }
            if (offset < memoryBytes.size) {
                val count: Int = min(len, memoryBytes.size - offset)
                System.arraycopy(memoryBytes, offset, buf, off, count)
                offset += count
                return count
            }
            return fileStream.read(buf, off, len)
        }

        @Throws(IOException::class)
        override fun close() {
            try {
                fileStream.close()
            } finally {
                Files.deleteIfExists(file)
            }
        }

    }

    companion object {
        const val MAX_MEMORY_SIZE: Int = 8 * 1024 * 1024
    }
}
